### PR TITLE
Restrict the decimal points of the quantity and price

### PR DESCRIPTION
**Pull request by ghost**
This PR solves issue #21. It addresses the need to restrict the number of decimal points for price and quantity fields, which are currently of type decimal but lack a limit on decimal places. This could result in values with an extremely large amount of decimal points to be sent. To resolve this, we implement constraints to match the requirements of the Exchange Platform EPEX, which permits only 1 decimal place for quantity and 2 decimal places for price.

It's important to note that even if we do not want to take the Exchange's restrictions into account when designing the API,, rounding values within the connector could lead to information loss from the user's perspective. For instance, if a user intends to order 0.15 MWh and the connector rounds to 0.1 MWh, it could cause significant discrepancies.

### CODE REVIEW

File: src/frequenz/client/electricity_trading/_client.py [hunk 1]
```diff
@@ -49,6 +50,25 @@ class _Sentinel:
...
+        name: The name of the value (for error messages).
+
+    Raises:
+        ValueError: If the value has more decimal places than allowed.
+    """
+    if value is not None:
+        if value.as_tuple().exponent < -decimal_places:  # type: ignore
+            raise ValueError(
+                f"The {name} cannot have more than {decimal_places} decimal places."
+            )
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/electricity_trading/_client.py]: I'd suggest making this function more bullet-proof with something like this (haven't tested it)
```python
    assert decimal_places >= 0, "The decimal places must be a non-negative integer."

    try:
        exponent = int(value.as_tuple().exponent)
        if abs(exponent) > decimal_places:
            raise ValueError(
                f"The {name} cannot have more than {decimal_places} decimal places."
            )
    except InvalidOperation as exc:
        raise ValueError(f"The {name} is not a valid decimal number.") from exc
```

Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py]: yep good call, fixed!

File: src/frequenz/client/electricity_trading/_client.py [hunk 2]
```diff
@@ -49,6 +50,32 @@ class _Sentinel:
...
+        decimal_places: The maximum allowed decimal places.
+        name: The name of the value (for error messages).
+
+    Raises:
+        ValueError: If the value has more decimal places than allowed.
+        ValueError: If the value is not a valid decimal number.
+    """
+    assert decimal_places >= 0, "The decimal places must be a non-negative integer."
+
+    if value is not None:
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/electricity_trading/_client.py]: The type-checker should alert us calling ` validate_decimal_places()` with `None` as the value parameter in the function only accepts a `Decimal` type. So if this function can be called with a None, then the signature should be `def validate_decimal_places(value: Decimal | None, decimal_places...`, otherwise this check is not needed.

**Review by daniel-zullo-frequenz — COMMENTED** [review]
Just a minor comment to check for and LGTM

File: src/frequenz/client/electricity_trading/_client.py [hunk 3]
```diff
@@ -49,6 +50,32 @@ class _Sentinel:
 NO_VALUE = _Sentinel()
 
 
+def validate_decimal_places(value: Decimal, decimal_places: int, name: str) -> None:
```
Comment (nhatcher-frequenz) [inline src/frequenz/client/electricity_trading/_client.py:55]: I think this is a very nice function to have some unit tests.

Comment (llucax) [inline src/frequenz/client/electricity_trading/_client.py:55]: I wouldn't make this function take a name, the caller should add this information when catching the exception, and if the exception is not caught, then the backtrace will give you this information.

File: src/frequenz/client/electricity_trading/_client.py (hunk 2 above)
Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py]: > The type-checker should alert us calling  validate_decimal_places() with None

oh yea right I forgot that I'm already checking it before calling the function my bad

File: src/frequenz/client/electricity_trading/_client.py [hunk 4]
```diff
@@ -49,6 +50,32 @@ class _Sentinel:
...
+
+    if value is not None:
+        try:
+            exponent = int(value.as_tuple().exponent)
+            if abs(exponent) > decimal_places:
+                raise ValueError(
+                    f"The {name} cannot have more than {decimal_places} decimal places."
+                )
+        except InvalidOperation as exc:
+            raise ValueError(f"The {name} is not a valid decimal number.") from exc
```
Comment (llucax) [inline src/frequenz/client/electricity_trading/_client.py]: I strongly recommend including the value you are validating here, like: `f"{value} cannot have more than {decimal_places} decimal places"`.

File: tests/test_client.py [hunk 5]
```diff
@@ -57,8 +57,8 @@ def set_up() -> Generator[Any, Any, Any]:
     )
     order_type = OrderType.LIMIT
     side = MarketSide.BUY
-    price = Price(amount=Decimal(50), currency=Currency.EUR)
-    quantity = Energy(mwh=Decimal(0.1))
+    price = Price(amount=Decimal("50"), currency=Currency.EUR)
+    quantity = Energy(mwh=Decimal("0.1"))
```
Comment (llucax) [inline tests/test_client.py:61]: Is this related to this PR?

**Review by llucax — COMMENTED** [review]
I don't know much about trading, but in terms of fault tolerance I wonder if providing some fixing of the value instead of refusing to work won't be more useful.

File: src/frequenz/client/electricity_trading/_client.py (hunk 3 above)
Comment (llucax) [reply src/frequenz/client/electricity_trading/_client.py:55]: On a second thought, if this function is not exported and it is just an internal utility function, then I think it is fine to keep the name.

File: tests/test_client.py (hunk 5 above)
Comment (ghost) [reply tests/test_client.py:61]: Yes because not having them in quotes resulted in the function to fail (converting floats to decimals like this seems to cause having e.g. 0.100000000000000000000000057901783 instead of 0.1).

Comment (llucax) [reply tests/test_client.py:61]: Ah, the niceties of floating point representation :D

File: src/frequenz/client/electricity_trading/_client.py (hunk 3 above)
Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py:55]: Indeed, I thought it would be too obvious at first but I guess it's still good to have

File: src/frequenz/client/electricity_trading/_client.py [hunk 6]
```diff
@@ -47,6 +48,35 @@ class _Sentinel:
...
+
+    try:
+        exponent = int(value.as_tuple().exponent)
+        if abs(exponent) > decimal_places:
+            raise ValueError(
+                f"The {name} cannot have more than {decimal_places} decimal places."
+            )
+    except InvalidOperation as exc:
+        raise ValueError(
+            f"{value} cannot have more than {decimal_places} decimal places"
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/electricity_trading/_client.py]: I think the error message here is not entirely right, maybe something like `f"The {value} for {name} is not a valid decimal number.`

File: src/frequenz/client/electricity_trading/_client.py [hunk 7]
```diff
@@ -47,6 +48,35 @@ class _Sentinel:
...
+        ValueError: If the value has more decimal places than allowed.
+        ValueError: If the value is not a valid decimal number.
+    """
+    assert decimal_places >= 0, "The decimal places must be a non-negative integer."
+
+    try:
+        exponent = int(value.as_tuple().exponent)
+        if abs(exponent) > decimal_places:
+            raise ValueError(
+                f"The {name} cannot have more than {decimal_places} decimal places."
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/electricity_trading/_client.py:74]:
```suggestion
                f"The {value} for {name} cannot have more than {decimal_places} decimal places."
```

File: src/frequenz/client/electricity_trading/_client.py [hunk 8]
```diff
@@ -47,6 +48,35 @@ class _Sentinel:
...
+    Validate that the decimal places of a given value do not exceed a specified limit.
+
+    Args:
+        value: The value to be checked.
+        decimal_places: The maximum allowed decimal places.
+        name: The name of the value (for error messages).
+
+    Raises:
+        ValueError: If the value has more decimal places than allowed.
+        ValueError: If the value is not a valid decimal number.
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/client/electricity_trading/_client.py]:
```suggestion
        ValueError: If the value has more decimal places than allowed
		    or the value is not a valid decimal number.
```

**Review by daniel-zullo-frequenz — COMMENTED** [review]
A few suggestions to check for given the latest changes

File: src/frequenz/client/electricity_trading/_client.py (hunk 7 above)
Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py:74]: Indeed, I'm getting distracted and not double checking, thanks! 🙏

File: tests/test_validate_decimal_places.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
```
Comment (daniel-zullo-frequenz) [inline tests/test_validate_decimal_places.py]: Nitpicking :)
```suggestion
# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
```

**Review by daniel-zullo-frequenz — DISMISSED** [review]
Only an optional cosmetic. LGTM

File: tests/test_validate_decimal_places.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+"""Tests for the function validating the amount of decimal places."""
+import unittest
+from decimal import Decimal
+
+from frequenz.client.electricity_trading._client import validate_decimal_places
+
+
+class TestValidateDecimalPlaces(unittest.TestCase):
```
Comment (nhatcher-frequenz) [inline tests/test_validate_decimal_places.py:11]: Thanks!

File: tests/test_validate_decimal_places.py (hunk 9 above)
Comment (ghost) [reply tests/test_validate_decimal_places.py]: hahah good catch 🕵️

**Review by daniel-zullo-frequenz — APPROVED** [review]
